app/helpers/locales_helper.rb

```diff
@@ -1,74 +1,80 @@
+    I18n.default_locale
   end
 
-  def detectable_locales
-    Loomio::I18n::DETECTABLE_LOCALES
-  end
-
-  def rtl_locales
-    Loomio::I18n::RTL_LOCALES
+  def supported_locales
+    YAML.load_file(Rails.root.join('config', 'locales.yml'))["supported_locales"]
   end
 
   def angular_locales
-    selectable_locales.map do |locale|
+    supported_locales.map do |locale|
       { key: locale, name: I18n.t(locale.to_sym, scope: :native_language_name) }
     end
   end
 
-  def set_application_locale
-    I18n.locale = best_locale
+  def save_detected_locale
+    if current_user.is_logged_in? && browser_detected_locales.any?
+      current_user.update_detected_locale(browser_detected_locales.first)
+    end
   end
 
-  def best_locale
-    (Array(locales_from_param)             | # locale from request param
-     Array(locales_from_user_preference)   | # locale from selected user preference
-     Array(locales_from_browser_detection) | # locales from browser headers
-     Array(locales_from_saved_browser_detection)   | # locale from selected user preference
-     Array(I18n.default_locale)).compact.first
+  private
+
+  def normalize(locale)
+    locale.to_s.sub('-','_')
   end
 
-  def save_detected_locale
-    if current_user.is_logged_in? && locales_from_browser_detection.any?
-      current_user.update_detected_locale(locales_from_browser_detection.first)
-    end
+  def strict_filter(locales)
+    Array(locales).compact.select do |locale|
+      supported_locales.select{|supported_locale| supported_locale == normalize(locale)}.first
+    end.first
   end
 
-  private
+  def fuzzy_filter(locales)
+    Array(locales).compact.map do |locale|
+      [normalize(locale),
+       strip_dialect(locale),
+       fallback_for(locale)].select do |version|
+        supported_locales.include? version
+      end.first
+    end.compact.first
+  end
 
-  def locales_from_param
-    filter_locales params[:locale], I18n.available_locales
+  def strip_dialect(locale)
+    locale.split('_').first
   end
 
-  def locales_from_user_preference
-    return unless current_user&.is_logged_in?
-    filter_locales(current_user.selected_locale, selectable_locales)
+  def fallback_for(locale)
+    fallbacks[locale] || fallbacks[strip_dialect(locale)]
   end
 
-  def locales_from_saved_browser_detection
-    return unless current_user&.is_logged_in?
-    filter_locales(current_user.detected_locale, selectable_locales)
+  def fallbacks
+    YAML.load_file(Rails.root.join('config', 'locales.yml'))["fallbacks"]
   end
 
-  def locales_from_browser_detection
-    parser = HttpAcceptLanguage::Parser.new(request.env["HTTP_ACCEPT_LANGUAGE"])
-    locales = parser.user_preferred_languages +
-              parser.user_preferred_languages.map { |l| l.split('-').first } # to catch locales like fr-fr, or en-nz
-    filter_locales(locales, detectable_locales)
+  def user_selected_locale
+    return nil unless current_user&.is_logged_in?
+    current_user.selected_locale
```

Comment:
current_user always returns something, so it won't be nil, it'll be a LoggedOutUser
So I think we'd rather just have LoggedOutUser respond to `selected_locale`? (I think it might do this already)

Then we can just called `current_user.selected_locale` instead of having this method

Reply:
Same for user_detected_locale; in fact it might make sense to just push down a method onto user (and Logged out user like `best_locale`

```
# user.rb
def best_locale
  selected_locale || detected_locale
end

# logged_out_user.rb
nil_methods :best_locale
```